stan/math/prim/meta/holder.hpp

```diff
@@ -51,110 +53,59 @@
+ * returned expression.
  */
 
-// This was implenmented following the tutorial on edding new expressions to
-// Eigen: https://eigen.tuxfamily.org/dox/TopicNewExpressionType.html
-
-namespace stan {
-namespace math {
-
-template <class ArgType, typename... Ptrs>
-class Holder;
-
-}  // namespace math
-}  // namespace stan
-
-namespace Eigen {
-namespace internal {
-
-template <class ArgType, typename... Ptrs>
-struct traits<stan::math::Holder<ArgType, Ptrs...>> {
-  typedef typename ArgType::StorageKind StorageKind;
-  typedef typename traits<ArgType>::XprKind XprKind;
-  typedef typename ArgType::StorageIndex StorageIndex;
-  typedef typename ArgType::Scalar Scalar;
-  enum {
-    // Possible flags are documented here:
-    // https://eigen.tuxfamily.org/dox/group__flags.html
-    Flags = (ArgType::Flags
-             & (RowMajorBit | LvalueBit | LinearAccessBit | DirectAccessBit
-                | PacketAccessBit | NoPreferredStorageOrderBit))
-            | NestByRefBit,
-    RowsAtCompileTime = ArgType::RowsAtCompileTime,
-    ColsAtCompileTime = ArgType::ColsAtCompileTime,
-    MaxRowsAtCompileTime = ArgType::MaxRowsAtCompileTime,
-    MaxColsAtCompileTime = ArgType::MaxColsAtCompileTime,
-    InnerStrideAtCompileTime = ArgType::InnerStrideAtCompileTime,
-    OuterStrideAtCompileTime = ArgType::OuterStrideAtCompileTime
-  };
-};
-
-}  // namespace internal
-}  // namespace Eigen
-
 namespace stan {
 namespace math {
 
-/**
- * A no-op Eigen operation. This object also owns pointers to dynamically
- * allocated objects used in its argument expression. When this object is
- * destructed, those objects are deleted.
- * @tparam Derived derived type
- * @tparam T type of the argument
- */
-template <typename ArgType, typename... Ptrs>
-class Holder
-    : public Eigen::internal::dense_xpr_base<Holder<ArgType, Ptrs...>>::type {
+template <typename BaseExpr, typename... Args>
+class Holder : private std::tuple<Args...>, public BaseExpr {
  public:
-  typedef typename Eigen::internal::ref_selector<Holder<ArgType, Ptrs...>>::type
-      Nested;
-  typename Eigen::internal::ref_selector<ArgType>::non_const_type m_arg;
-  std::tuple<std::unique_ptr<Ptrs>...> m_unique_ptrs;
-
-  explicit Holder(ArgType&& arg, Ptrs*... pointers)
-      : m_arg(arg), m_unique_ptrs(std::unique_ptr<Ptrs>(pointers)...) {}
+  using BaseExpr::BaseExpr;
+  using BaseExpr::operator=;
+  /**
+   * Constructor
+   * @param f
+   * @param args
+   */
+  template <typename Functor, require_same_t<decltype(std::declval<Functor>()(
+                                                 std::declval<Args&>()...)),
+                                             BaseExpr>* = nullptr>
+  explicit Holder(const Functor& f, Args&&... args)
+      : std::tuple<Args...>(std::forward<Args>(args)...),
+        BaseExpr(index_apply<sizeof...(Args)>([this, &f](auto... Is) {
+          return f(std::get<Is>(*static_cast<std::tuple<Args...>*>(this))...);
+        })) {}
```

Comment:
Just to make sure I understand, for something like 

```cpp
template<typename T>
auto f(const T& x){
   const Eigen::Ref<const Eigen::VectorXd>& x_ref = x;
  make_holder([](const auto& x) { return x.cwiseProduct(x);}, x_ref);
}
```

This constructor is called which stores the `x_ref` by passing it to the inherited tuples constructor, then the lambda is invoked on `x_ref` with the expression stored in `BaseExpr`. Is that correct? It might be good to include that in the docs as I think in the past we've been confused about how this mechanic works.